### ISSUE TITLE
[question] How to create local component?

### ISSUE THREAD

**Issue by xkwangcn** [redhat-developer/vscode-openshift-tools#2128]
When I use Watch commands for my pushed component, there is an info says: Watch is supported only for Components with local or binary source type.

My question is what local components means? and how can I create it?

Thank you very much.

**Comment by xkwangcn** [comment]
When I use Debug command, there is also a similar info says: Debug is supported only for local components.

**Comment by mohitsuman** [comment]
Hello @xkwangcn , thanks for reaching out. For your query, here are the some of the required steps.

1. Local Component -  Components are divided into 3 different sections: _local_, _binary_ and _git_ component. 

Now a `local component` is something which is created using local directory(workspace) as a source for the Component. Say you have a directory cloned and added in your vscode workspace. Now this directory will be used to create a component. 
Debug and watch features are ONLY available for these workspaces. Hence you see those info messages.

How to create a local component ?
1. Add a directory/project into your vscode workspace
2. Open the OpenShift Extension Application Explorer and click on `New Component`
<img width="536" alt="Screenshot 2021-05-11 at 2 10 18 PM" src="https://user-images.githubusercontent.com/3133356/117785914-a6359280-b262-11eb-9717-bbb8efd5ec67.png">

3. This will open a prompt to provide 
- Project Name
- Application Name
- Provide the context folder : This is where you have the select the workspace directory you added in Step 1
- Provide Component Name

Now the component will be created and it will be a local component (as it has the source of your local workspace). Going forward once the component is pushed, watch and debug actions can be performed.

**Comment by xkwangcn** [comment]
@mohitsuman Thanks for your replay. It's helpful.
but when I click watch, I didn't see some logs? only here a quick message appear and disappear? 
<img width="511" alt="image" src="https://user-images.githubusercontent.com/29972855/117789862-58795400-b27b-11eb-8418-655fb2b23c2f.png">
and do you know the reason? thank you.

**Comment by mohitsuman** [comment]
> about debug, I don't know how to do next? could you help about this steps? thank you very much.
> <img alt="image" width="1457" src="https://user-images.githubusercontent.com/29972855/117790330-d50c3280-b27b-11eb-81f8-aa696b1198e0.png">

For debug functionality, we have a shirt video which showcases to debug a nodejs component. This should be helpful https://github.com/redhat-developer/vscode-openshift-tools#debug-nodejs-component

**Comment by xkwangcn** [comment]
> @mohitsuman Thanks for your replay. It's helpful.
> but when I click watch, I didn't see some logs? only here a quick message appear and disappear?
> <img alt="image" width="511" src="https://user-images.githubusercontent.com/29972855/117789862-58795400-b27b-11eb-8418-655fb2b23c2f.png">
> and do you know the reason? thank you.

@mohitsuman hi, is there guide about watch? I don't know how to ensure the watch command correct in my environment. Thank you.

**Comment by dgolovin** [comment]
@xkwangcn There is no guide for the extension. Watch is a feature that automatically detects changes to local code and applies the changes to the component deployed (pushed) to cluster in real time. You seem to hit an issue. I can replicate it and reported here [#2137](https://github.com/redhat-developer/vscode-openshift-tools/issues/2137).

**Comment by xkwangcn** [comment]
@dgolovin Thanks for your replay.
I will describe my specific steps:
1. I created a local component named localnodejs
<img width="345" alt="image" src="https://user-images.githubusercontent.com/29972855/118225404-a37fab00-b4b7-11eb-9413-56382a5cba34.png">
2. Then I click watch 
<img width="341" alt="image" src="https://user-images.githubusercontent.com/29972855/118225490-d4f87680-b4b7-11eb-9d3a-b717928a7e6d.png">
3. as the above image shows, I can see nothing and no logs in the watch sessions. 

My question is how to debug and solve this problem? Thank you very much.

**Comment by dgolovin** [comment]
@xkwangcn the issue is fixed and we aiming to release next week.

**Comment by kandarpamalipeddi** [comment]
@dgolovin 

I tried debug feature as mentioned in the GIF at : https://github.com/redhat-developer/vscode-openshift-tools#debug-nodejs-component

But after selecting the debug option, it is stuck at
![image](https://user-images.githubusercontent.com/69784217/122228417-728aff80-ced5-11eb-9df5-fb6fc7f175f0.png)

**Comment by dgolovin** [comment]
@kandarpamalipeddi It supposed to fail on timeout, but I guess it does not in your case.
Unfortunately, we have not implemented cancellation for long running commands yet #2098.

I assume you are using nodejs component. Would you do the following steps to understand what is going on:
1. Run 'Describe' command for component you are debugging (it should be running in terminal view)
2. Modify original command `"/full/path/to/odo" describe` to  `"/full/path/to/odo" debug port-forward` and run it in the same terminal
3. Post output here

**Comment by kandarpamalipeddi** [comment]
1. Output of `odo describe` 
```
"c:\Users\KandarpaMalipeddi\.vscode\extensions\redhat.vscode-openshift-connector-0.2.7\out\tools\win32\odo.exe" describe
W0617 10:45:00.589740   13800 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
Component Name: nodejs
Type: nodejs
Environment Variables:
 · PROJECTS_ROOT=/project
 · PROJECT_SOURCE=/project
 · DEBUG_PORT=5858
URLs:
 · http://http-3000-nodejs-debug1.apps.shivanij-3-47.ibm.com exposed via 3000
```
2. `debug port-forward` output:
```
"c:\Users\KandarpaMalipeddi\.vscode\extensions\redhat.vscode-openshift-connector-0.2.7\out\tools\win32\odo.exe" debug port-forward
Started port forwarding at ports - 5858:5858
```
Its waiting after this statement.

**Comment by dgolovin** [comment]
Extension parses `Started port forwarding at ports - 5858:5858` and extract left side (local port) of port mapping `5858:5858`. Then tries waits for `localhost:5858` to open and start debugging session on it.

My guess the 'localhost' cannot be resolved and that leads to never ending wait for port to be open.
Do you have record for localhost in C:\WINDOWS\System32\drivers\etc\hosts? Something like 
```
127.0.0.1 localhost
```

**Comment by kandarpamalipeddi** [comment]
localhost is associated with 127.0.0.1, Please see below
![image](https://user-images.githubusercontent.com/69784217/122419315-72a7ff80-cfa8-11eb-92b2-1bcf53f45d6d.png)

Getting below error:
```
"c:\Users\KandarpaMalipeddi\.vscode\extensions\redhat.vscode-openshift-connector-0.2.7\out\tools\win32\odo.exe" debug port-forward
Started port forwarding at ports - 5858:5858
E0617 20:08:25.329125   23660 portforward.go:233] lost connection to pod
```

**Comment by jeffmaury** [comment]
That may be caused by the pod being killed on the cluster maybe checking events on the cluster could help

**Comment by dgolovin** [comment]
@kadel there is no way to recover lost connection even if pod getting restored. Am I right? odo should just exit in the case it is loosing connection to pod, should it not?

**Comment by dgolovin** [comment]
@kandarpamalipeddi thank you for

**Comment by dgolovin** [comment]
Extension could do better job here:
1. There should have Cancel option for every long running command or timeout should be set to prevent commands for running (see the issue I mentioned before #2098)
2. Every long running command should give access to stdout/stdin output for debugging purposes
@kandarpamalipeddi thank you for your feedback on the issue

**Comment by kandarpamalipeddi** [comment]
@dgolovin 
Any updates on debug issue on Power ?
Do you need any info from me ?

Regards,
Kandarpa

**Comment by dgolovin** [comment]
@kandarpamalipeddi sorry for delay. Nothing has been done yet.

Can you check what is going on cluster side? What is the reason behind pod termination? Is it remote or local single node cluster?

**Comment by kandarpamalipeddi** [comment]
@dgolovin 
I am trying on a remote cluster (Its a PPC64LE cluster).
Will check and give you more details about cluster side logs.

**Comment by sjagadale** [comment]
@dgolovin 

**We found following log in pod, we are investigating this.** 

```
time="2021-08-06T09:42:35Z" level=debug msg="no auth required"
time="2021-08-06T09:42:35Z" level=info msg="stop the program" program=devrun
time="2021-08-06T09:42:35Z" level=info msg="force to kill the program" program=devrun
time="2021-08-06T09:42:35Z" level=error msg="program stopped with error:signal: killed" program=devrun
time="2021-08-06T09:42:35Z" level=info msg="Stopped by user, don't start it again" program=devrun
time="2021-08-06T09:42:36Z" level=info msg="stop the program" program=debugrun
time="2021-08-06T09:42:36Z" level=info msg="force to kill the program" program=debugrun
time="2021-08-06T09:42:39Z" level=debug msg="no auth required"
time="2021-08-06T09:42:39Z" level=debug msg="succeed to find process:debugrun"
time="2021-08-06T09:42:39Z" level=info msg="try to start program" program=debugrun
time="2021-08-06T09:42:39Z" level=info msg="success to start program" program=debugrun
ODO_COMMAND_DEBUG is npm run debug
Changing directory to /project
Executing command cd /project && npm run debug
npm ERR! missing script: debug

 

npm ERR! A complete log of this run can be found in:
npm ERR!     /opt/app-root/src/.npm/_logs/2021-08-06T09_42_39_776Z-debug.log
time="2021-08-06T09:42:40Z" level=debug msg="wait program exit" program=debugrun
time="2021-08-06T09:42:40Z" level=error msg="program stopped with error:exit status 1" program=debugrun
time="2021-08-06T09:42:40Z" level=info msg="Don't start the stopped program because its retry times 0 is greater than start retries 0" program=debugrun
time="2021-08-06T09:42:43Z" level=debug msg="no auth required"
```

**Comment by sjagadale** [comment]
@dgolovin
We have done below analysis. 
1.	Windows OS to Power Cluster – Debug feature hangs/failed . 
2.	MacOS to Power cluster – Debug Feature is successful. 
We are in process of testing one more scenario then we can conclude if its problem with OpenShift Connector on Windows or not.  On Windows and MacOS OpenShift Connector extension version is v0.2.9

**Comment by sjagadale** [comment]
@dgolovin 
Attaching screen shots of Mac Book VS code to Power OpenShift cluster  Debug working condition. 

![VS to power cluster ](https://user-images.githubusercontent.com/86092461/129674588-68d33fad-79d4-45ed-9411-a3b8fa32a108.png)

![InkedVS to power cluster 2_LI](https://user-images.githubusercontent.com/86092461/129674943-85fc155e-e607-41c8-9197-357e779116e8.jpg)

**Comment by sjagadale** [comment]
for further debug option we have created new issue  - https://github.com/redhat-developer/vscode-openshift-tools/issues/2229

@dgolovin

**Comment by dgolovin** [comment]
#2229 was fixed in release v0.2.11. Closing this issue.
